### ISSUE TITLE
Refactoring support

### ISSUE THREAD

**Issue by spyder-issues-migrator** [spyder-ide/spyder#415]
_From [joan.man...@gmail.com](https://code.google.com/u/105883090728463490757/) on 2010-10-14T06:03:28Z_

Rope can provide a great refactoring support, I note that spyder use it but not for refactoring (masive and safe renaming).

Spyder would provide a great feature if you implement a good support for refactoring.

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=415_

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-10-26T04:34:50Z_

**Status:** Accepted  
  **Labels:** -Type-Defect -Priority-Medium Type-Enhancement Priority-Low

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-05-16T01:14:51Z_

**Labels:** Cat-Editor

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-05-18T00:30:15Z_

**Status:** ContributorNeeded

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2012-03-18T14:11:22Z_

**Status:** HelpNeeded  
  **Labels:** -Type-Enhancement Type-Enh

**Comment by yw5aj** [comment]
Had a couple of friends who switched to PyCharm simply due to the refactoring capability... Felt that this is a really needed feature.

**Comment by Nodd** [comment]
Really needed features really need contributors ! ;)
I agree that it would be great to have, thanks for the reminder.

**Comment by bmer** [comment]
How can I implement this in Spyder? I am really finding that I want this too

**Comment by CAM-Gerlach** [comment]
General note: #1937 is a near duplicate of this one; the nominal scope in that issue's OP is more limited (variable renaming, etc), but while what gets described in the comments (a full refactoring pane, etc.) covers what is being requested here.

**Comment by cknoll** [comment]
This issue has the tag community:Help wanted. I am considering to contribute on this because I would really appreciate that feature but some hints on were to start would be helpful.

**Comment by ccordoba12** [comment]
@cknoll, thanks for offering help. We will address this after PR #4751 is merged, sorry for the confusion.

**Comment by f4ww4z** [comment]
Any updates on this? Would love to see this feature soon. @ccordoba12

**Comment by ccordoba12** [comment]
@andfoy, most probably we'll leave this for Spyder 5, to be released in 2020.

**Comment by DBerke** [comment]
I'd like to voice some support for this feature; I've been using Spyder for somewhat over a year now, and this is the _**one**_ feature I still really miss from other editors. (It's like syntax highlighting: once you've used it it's hard to do without it!) I see you've removed the "community help wanted" label some time ago, but I'd be willing to work on implementing it myself if that would help it get in faster.

**Comment by Kischy** [comment]
This feature would be of great help!

**Comment by danuker** [comment]
@DBerke, @Kischy: it seems there is pull request #2964 about this, and what it needs to do is use `pyls`. Would any one of you like to continue the work there?

**Comment by jitseniesen** [comment]
> it seems there is pull request #2964 about this, and what it needs to do is use `pyls`. 

As the author of that pull request, I came to the conclusion that the rope library (which I was using to do the actual refactoring) is not a good fit because the rope library works on files save on disc, while we want to act on files in the editor. Thus, you need either to make substantial changes to rope (which does not seem to be very active) or find another library to do the refactoring. I don't know what is the best approach, but I have no plans to work on it myself any time soon.

@ccordoba12 Perhaps we should add the community help wanted tag back in now that support for PyLS is merged?

**Comment by danuker** [comment]
> we want to act on files in the editor

For me it would be OK to require saving all files before refactoring. What do you think?

**Comment by sten0** [comment]
Here are two alternatives to Rope:
[Bowler](https://github.com/facebookincubator/bowler) (uses [fissix](https://github.com/jreese/fissix), a backported lib2to3 that is intended to enable Bowler to always support the latest Python grammar)
  * Pro: Allegedly much easier to work with than Rope.
  * Pro: Living upstream.
  * Fissix is bleeding edge.  That's either a pro or a con, depending on one's perspective.  It's also being used to provide addition QA for Python upstream lib2to3.

[RedBaron](https://github.com/PyCQA/redbaron) (Supports up to Python 3.7 grammar)
  * From the upstream description **"For now, RedBaron can be considered in alpha, the core is quite stable but it is not battle tested yet and is still a bit rough"**

**Comment by texadactyl** [comment]
I would like this feature very much!  I found it very useful in Eclipse.  However, I tended to use it to rename a specific variable inside a single file.  Special case, perhaps.

Hopefully, scope will be respected.  
E.g.
```
squared = lambda x: x * x
def halved(x):
     return x / 2.0

x = 3
x += 42
```
If I refactor x in the outer scope to puppy, this will be the result:
```
squared = lambda x: x * x
def halved(x):
     return x / 2.0

puppy = 3
puppy = += 42
```
I am sticking with Spyder whether or not this enhancement request is addressed.  It catches my worst mistakes.

**Comment by goanpeca** [comment]
>I am sticking with Spyder whether or not this enhancement request is addressed. It catches my worst mistakes, makes my overall Python development easier than it would be otherwise, especially converting my Python 2 sources to Python 3.

Glad to hear that! We are working on refactoring support so it will come in some months ;-)


Thanks for using spyder !

**Comment by texadactyl** [comment]
@goanpeca  As a developer, I can directly relate trying to balance personal time/energy with work, "hobby time" projects, and family.  It is not easy.

**Comment by goanpeca** [comment]
We will add refactoring support via the python-language-server (now integral part of Spyder 4.0.0 released yesterday!) so you can also follow https://github.com/palantir/python-language-server to be on the loop. We also help develop that package!

**Comment by texadactyl** [comment]
Limited refactoring that would be useful:
- inside the scope of one specific "def" or "lambda".
- between source file line numbers

**Comment by ccordoba12** [comment]
@goanpeca, I think refactoring is already implemented in the PyLS. We just need to provide a UI in our side to perform it.

**Comment by goanpeca** [comment]
> I think refactoring is already implemented in the PyLS. We just need to provide a UI in our side to perform it.

Nice :-p

**Comment by Lukious** [comment]
The refactoring function of pyLS is not perfect.
 Is it not possible to support refactoring the formal spyder version?

**Comment by real-yfprojects** [comment]
Doesn't  `jedi` allow simple refactoring too?

**Comment by rhkarls** [comment]
> 
> 
> Doesn't `jedi` allow simple refactoring too?

It seems like it does, but that `rope` has more features. From what I gather, pyLS uses `rope` for this but I'm not sure to what degree the features are implemented in pyLS. Anyway, would be a great improvement to spyder to have basic refactoring support, like renaming!

**Comment by bherwig** [comment]
Now and then I check out this issue .... always going back to start pycharm. :-/ (And regarding the pyLS-thing ..... well, mh, palantir? Really? :( Improving science at what cost and with the company of what kind of researchers? Please use rope.)

**Comment by rhkarls** [comment]
I came across https://github.com/python-rope/pylsp-rope - is this something that can potentially be used with Spyder?

**Comment by ccordoba12** [comment]
Yep, that's the perfect fit to provide refactoring. Unfortunately, we don't have much time right now to implement refactoring at the moment.

**Comment by Greedquest** [comment]
@ccordoba12 I'm slightly confused - in the readme of `python-lsp/python-lsp-server` it says:

> If the respective dependencies are found, the following optional providers will be enabled:
> - [Rope](https://github.com/python-rope/rope) for Completions and renaming

Is this just outdated documentation left over when the repo was forked from https://github.com/palantir/python-language-server to https://github.com/python-lsp/python-lsp-server? 

---

Or is it this:
> @goanpeca, I think refactoring is already implemented in the PyLS. We just need to provide a UI in our side to perform it.

The language server used by Spyder *can* do the refactoring but there is no GUI element to access it? If so then how does a 3rd party plugin like `pylsp-rope` manage to add itself to the GUI (or does it also not do anything in Spyder, just in other IDEs that run on an lsp)? Sorry, just a bit confused about what needs to be done to support refactoring (particularly renaming) or if it's already partially supported in Spyder through plugins/ extras to the lsp?

**Comment by ccordoba12** [comment]
> The language server used by Spyder can do the refactoring but there is no GUI element to access it?

Yep, this is it.

> If so then how does a 3rd party plugin like pylsp-rope manage to add itself to the GUI

There's no way for that plugin to register graphical elements to allow Spyder users to do refactoring with it. Editors or IDEs (like Spyder or VSCode) have to provide that, which behind the scenes sends a refactoring request to the Python language server (PyLSP). That request is processed by pylsp-rope, its result is sent back to the graphical interface, which finally handles it and displays the result appropriately.

That's how completion, linting, folding and a lot of other features are implemented in Spyder right now.

**Comment by rswitt** [comment]
I searched to see whether refactoring support in Spyder has already been suggested, and I see that it has.  I understand that priorities must be managed.

However, I will say, that I know a good number of people that choose to use PyCharm *solely* due to refactoring support, so you might want to consider that for future builds.

**Comment by ccordoba12** [comment]
Hey @rswitt, thanks for the input. Unfortunately we don't have the resources to implement this right now (we're a self-funded project, not affiliated to Anaconda or other institutions).
